orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+    const std::string& tensor_name = property_proto.name();
+    auto data_type = property_proto.data_type();
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto::FLOAT: {
+        const std::vector<float>& flt_parsed = ONNX_NAMESPACE::ParseData<float>(&property_proto);
+        ORT_ENFORCE(flt_parsed.size() == static_cast<size_t>(1), "only support scalar float properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<float>>(
+            tensor_name, flt_parsed.at(0)));
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto::STRING: {
+        const std::vector<std::string>& str_parsed = ParseStringData(&property_proto);
+        ORT_ENFORCE(str_parsed.size() == static_cast<size_t>(1), "only support scalar string properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<std::string>>(
+            tensor_name, str_parsed.at(0)));
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto::INT64: {
+        const std::vector<int64_t>& int_parsed = ONNX_NAMESPACE::ParseData<int64_t>(&property_proto);
+        ORT_ENFORCE(int_parsed.size() == static_cast<size_t>(1), "only support scalar int64_t properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<int64_t>>(
+            tensor_name, int_parsed.at(0)));
+        break;
+      }
+      default:
+        ORT_THROW("Unsupported input data type of ", data_type);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtLoadModuleStatesInternal(
+    const PathString& parameter_folder_path, ModuleCheckpointStates& module_states) {
+  // Parameter parsing.
+  const PathString module_state_file_path = GetTensorProtoFilePath(parameter_folder_path, k_param_root_prefix);
+  auto& named_parameters = module_states.named_parameters;
+  std::vector<ONNX_NAMESPACE::TensorProto> param_tensor_protos{};
+  auto file_read_status = WithOpenFile(
+      module_state_file_path, true,
+      [&param_tensor_protos](int fd) {
+        google::protobuf::io::FileInputStream input{fd};
+        ORT_RETURN_IF_ERROR(ReadProtoMessageSequence(param_tensor_protos, input));
+        return Status::OK();
+      });
+
+  if (!file_read_status.IsOK()) {
+    LOGS_DEFAULT(WARNING) << ToUTF8String(module_state_file_path) << " module state file read failure, skip it.";
+    return Status::OK();
+  }
+
+  std::vector<const ONNX_NAMESPACE::TensorProto*> param_tensor_proto_ptrs{};
+  for (ONNX_NAMESPACE::TensorProto& param_tensor_proto : param_tensor_protos) {
+    param_tensor_proto_ptrs.emplace_back(&param_tensor_proto);
+  }
+
+  std::unordered_map<std::string, OrtValue> name_to_ort_values;
+  ORT_RETURN_IF_ERROR(CreateOrtValuesFromTensorProtos(param_tensor_proto_ptrs, name_to_ort_values));
+  for (auto it = name_to_ort_values.begin(); it != name_to_ort_values.end(); ++it) {
+    named_parameters.insert({it->first, std::make_shared<Parameter>(it->first, it->second)});
+  }
```

Comment:
in the resume from checkpoint case, it is possible that we will occupy double the memory in terms of parameters since we are not replacing the Module and optimizer's parameters. Is there a way to just swap out the OrtValues instead? This way the gradients don't need to be reallocated every time we load checkpoint

Reply:
Sorry I did not catch the point. In the initial working flow designed for on-device, a checkpoint is loaded first, and then use the value to construct Module and Optimizer. So what do you mean of "double the memory"? 

In normal working flow (I added in the on-device doc as design alternatives), Module and Optimizer is created first, then Module.LoadStateDict(loaded_checkpoint_states)/Optimizer.LoadStateDict(loaded_checkpoint_states). When we want to support this way, yes, there will be double parameters or parameter states during the LoadStateDict execution. Are you referring to this?

Reply:
If it is the case I mention, we can optimizer this once Module and Optimizer implementation is done.